### DISCUSSION TITLE
'Cannot perform I/O on behalf of a different request' errors when running GraphQL queries out of edge middleware

### DISCUSSION THREAD

**Discussion by art-alexeyenko** [vercel/vercel#8668] [Help] [answered]
We have a nextjs website deployed to Vercel and we observe this error (occasionally, not on every request but often) when multiple users access a page:
`Error: Cannot perform I/O on behalf of a different request. I/O objects (such as streams, request/response bodies, and others) created in the context of one request handler cannot be accessed from a different request's handler. This is a limitation of Cloudflare Workers which allows us to improve overall performance.`

It fails when sending a GraphQL query via this method:
`async request<T>(
    query: string | DocumentNode,
    variables?: { [key: string]: unknown }
  ): Promise<T> {
    return new Promise((resolve, reject) => {
     // only logging and setting up timeout here

      this.client
        .request(query, variables)
        .then((data: T) => {
          clearTimeout(abortTimeout);
          this.debug('response: %o', data);
          resolve(data);
        })
        .catch((error: ClientError) => {
          this.debug('response error: %o', error.response || error.message || error);
          reject(error);
        });
    });`

this.client has a Client type from 'graphql-request' npm module.
The request method itself is called with await.

I've looked for similar cases but came out with nothing, it seems like some quirk of Cloudflare, but perhaps there's something wrong with the code?

Are there any specific considerations about sending outbound requests from edge middleware?
Are there any considerations in using '.then()' format versus 'await' when resolving promises in middleware?

**Comment by amyegan** [comment]
There are some [Edge Middleware limitations](https://vercel.com/docs/concepts/functions/edge-middleware/limitations#) which could cause errors. Edge Middleware [cannot produce a response body](https://nextjs.org/docs/messages/middleware-upgrade-guide#explanation-1) or a runtime error will be thrown.

I don't think I'm seeing the full picture of what you need to accomplish. Can you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) or overview description of the flow through your app? That will help me better understand your use case so I can be more helpful for you.

**Reply by art-alexeyenko** [reply]
For sure! I don't think we have a reproduce example (I'll try to setup one), but I can outline the flow of the app.
nextjs app uses middleware which loads plugins to be executed one by one:
```
export default async function middleware(
  req: NextRequest,
  ev: NextFetchEvent
): Promise<NextResponse> {
  const response = NextResponse.next();

  return (Object.values(plugins) as MiddlewarePlugin[])
    .sort((p1, p2) => p1.order - p2.order)
    .reduce((p, plugin) => p.then((res) => plugin.exec(req, res, ev)), Promise.resolve(response));
}
```

Plugins are being imported from a subfolder. Issue happens even if only one plugin is used.
Plugin would look like this:
```
class RedirectsPlugin implements MiddlewarePlugin {
  private redirectsMiddleware: RedirectsMiddleware;
  order = 0;

  constructor() {
    this.redirectsMiddleware = new RedirectsMiddleware();
  }

  async exec(req: NextRequest): Promise<NextResponse> {
    return this.redirectsMiddleware.getHandler()(req);
  }
}

export const redirectsPlugin = new RedirectsPlugin();
```

This particular plugin retrieves data (redirects map) from a GraphQL endpoint and rewrites/redirects NextResponse if needed. The error is thrown before we get to that last part however, 


RedirectsMiddleware class is used like this:
```
public **getHandler**(): (req: NextRequest) => Promise<NextResponse> {
    return this.handler;
  }

  private **handler** = async (req: NextRequest): Promise<NextResponse> => {
    // Find the redirect from result of RedirectService
    const **existsRedirect** = await this.**getExistsRedirect**(req); **// this is the start of the call chain that throws**

    if (!existsRedirect) {
      return NextResponse.next();
    }
```

this.**getExistsRedirect** is the start of the call chain to get data from remote GraphQL endpoint. I will outline the chain below:
```
private async **getExistsRedirect**(req: NextRequest): Promise<RedirectInfo | undefined> {
    const redirects = await this.redirectsService.fetchRedirects(); **// this fails**
```

fetchRedirects from redirect service looks like:

```
/**
   * Fetch an array of redirects from API
   * @returns Promise<RedirectInfo[]>
   * @throws {Error} if the siteName is empty.
   */
  async fetchRedirects(): Promise<RedirectInfo[]> {
    const siteName: string = this.options.siteName;

    if (!siteName) {
      throw new Error('siteNameError');
    }

    const redirectsResult: Promise<RedirectsQueryResult> = this.graphQLClient.request(this.query, {
      siteName,
    }); **// this fails**

    return redirectsResult
      .then((result: RedirectsQueryResult) => result?.site?.siteInfo?.redirects || [])
      .catch((e) => Promise.reject(e));
  }
```

and finally the request method implementation is in my original post.

This is kinda long-winded, so I'll try to setup a repro sample with graphQL mock too. Anyway, hope this helps a bit.

**Reply by art-alexeyenko** [reply]
RE: reproducible example, please check here: https://github.com/art-alexeyenko/lock-stock-mock
Steps to reproduce:
1) Host mock-express in vercel
2) host next-mid-ts as a separate vercel site
3) set mock's URL either in lib\middleware\plugins\redirects.ts :
endpoint: process.env.GRAPHQL_ENDPOINT || "https://lock-stock-mock.vercel.app/api/graphql/v1",

or via GRAPHQL_ENDPOINT env variable, with the '/api/graphql/v1' path.

4) Check function logs after next-mid-ts is deployed.
OR
Check function logs after requesting the home page a few times, from different sessions.
https://www.screencast.com/t/RRt5BsuWz

The issue is intermittent. The most reliable way to see it is to open pages that use middleware (i.e. homepage) from different machines simultaneously.
Also please nevermind the graphql error - the I/O error would occur regardless of mock sending correct or incorrect gql response.

**Comment by art-alexeyenko** [comment, marked as answer]
To anyone who may stumble on this: the root cause is AbortController usage in fetch calls.
You may be initializing client from graphql-request like this:
```
new Client(endpoint, {
      headers: this.headers,
      fetch: clientConfig.fetch,
      signal: this.abortController.signal as RequestInit['signal'],
    });
```

or using a fetch call like this:
```
fetch(this.endpoint, {
        method: 'POST',
        signal: thisabortController.signal,
        headers: headers,
        body: body
       ),
      }).then(....)

```
but this will fail (occasionally) when executed in Edge Middleware.

Both approaches should work fine if signal is not provided.

**Reply by ambrauer** [reply]
Not really a fix if you actually need to use an AbortSignal to cancel the request (e.g. for timeout functionality). This is quite crucial especially in middleware where you may need this control for performance reasons.

**Comment by art-alexeyenko** [comment]
Issue registered here : https://github.com/vercel/edge-runtime/issues/169
